workshops/shared_clipboard/README.md

```diff
@@ -0,0 +1,1242 @@
+- We handle errors
+- If `localClipboard` is the same as the local clipboard, we wait for some time and check again. If it is still the same, we sleep again. This cycle continues *until* the clipboard changes. When it does, `localClipboard` won't be the same as the actual clipboard on the computer so we'll exit the loop. Because we exited the loop, the code will send the clipboard! 
+
+When the function is first called, it immediately sends the current clipboard. Why would we want this? This is because when you join a clipboard, you probably want the other computers to sync up immediately.
+
+Right now, we check every second because we set `secondsBetweenChecksForClipChange` to 1 when declaring variables. You can change it!
+
+You: "You have ANOTHER function to send the clipboard to a writer??"
+
+At this point, you might be like "Oh my god so many functions it never ends" but in the end it's more readable, it's neater, it's easier to understand what each function is doing and it's easier to manage. A few minutes of looking at the entire source code will convince you of this!
+
+### Send the clipboard
+
+The reason we have a seperate function for sending the clipboard is so both the server and the client can send clipboards to each other and we don't have to use duplicate code.
+
+```go
+func sendClipboard(w *bufio.Writer, clipboard string) error {
+
+}
+```
+
+This function will send the clipboard to a writer and return an error if anything goes wrong. We return the error so that we can check for clients that are unreachable and if they are, we can remove them from the list of clients.
+
+When we're sending the clipboard, we put "STARTCLIPBOARD" over it and "ENDCLIPBOARD" below it so that when we're parsing the clipboard the parser knows where one clipboard ends and where another starts (we'll be sending multiple clipboards because they can change, you know). Add in:
+```go
+var err error
+clipString := "STARTCLIPBOARD\n" + clipboard + "\nENDCLIPBOARD\n"
+```
+Sometimes the code gets an empty clipboard for some reason (I have no idea why. Please do tell me if you find out why. I'm **@Ishan** on Slack) so add this to check if it is empty and to not send it if it is empty so that the other computers don't get their clipboards emptied (tongue twister, eh?):
+```go
+if clipboard == "" {
+    debug("was going to send empty string but skipping")
+    return nil
+}
+```
+
+Now we need to show the user what's being sent and then actually send the clipboard.
+Add in:
+```go
+debug("sent:", clipboard)
+_, err = w.WriteString(clipString)
+if err != nil {
+    return err
+}
+err = w.Flush()
+return err
+```
+Because this is a buffered writer, we have to remember to flush. (flush sends all the data in the buffer down the- I mean to the computer at the other end.) Along the way we check for errors and if we find any, we return them.
+
+### Check for sent clipboards
+This is going to be the largest function:
```

Comment:
Yep that's just to show it all together. There's so many different for loops nested inside each other that it would be hard to put it all together in little chunks. Instead, I explain each block after the code fences. Let me know what I should do.

Reply:
sure, thanks!